<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNA-seq analysis in R • RNAseqRworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="RNA-seq analysis in R">
<meta property="og:description" content="RNAseqRworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAseqRworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/00-r-rstudio-intro.html">R for RNA-seq</a>
</li>
<li>
  <a href="../articles/06-rnaseq-day1.html">Differential expression analysis</a>
</li>
<li>
  <a href="../articles/08-applying-rnaseq.html">Applying RNA-seq</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://mblue9.github.io/RNAseq-R_testing">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/mblue9/RNAseq-R_testing/">
    <span class="fab fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="00-r-rstudio-intro_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNA-seq analysis in R</h1>
            <h3 class="subtitle">R for RNAseq</h3>
                        <h4 class="author">Harriet Dashnow</h4>
            
            <h4 class="date">27 April 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mblue9/RNAseq-R/blob/master/vignettes/00-r-rstudio-intro.Rmd"><code>vignettes/00-r-rstudio-intro.Rmd</code></a></small>
      <div class="hidden name"><code>00-r-rstudio-intro.Rmd</code></div>

    </div>

    
    
<p>Here we cover some R skills that are particularly useful for RNAseq analysis, for example data frame manipulation, factors and subsetting using logical statements. If you are a more advanced R user, you may not need this.</p>
<object data="http://www.africau.edu/images/default/sample.pdf" type="application/pdf" width="700px" height="700px">
<embed src="http://www.africau.edu/images/default/sample.pdf"><p>
This browser does not support PDFs. Please download the PDF to view it: <a href="http://www.africau.edu/images/default/sample.pdf">Download PDF</a>.
</p>
</embed></object>
<object data="../slides/RNASeq_basics.pdf" type="application/pdf" width="700px" height="700px">
<embed src="../slides/RNASeq_basics.pdf"><p>
This browser does not support PDFs. Please download the PDF to view it: <a href="http://www.africau.edu/images/default/sample.pdf">Download PDF</a>.
</p>
</embed></object>
<div id="acknowlegements" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowlegements" class="anchor"></a>Acknowlegements</h2>
<p>Some of this material was adapted from <a href="https://github.com/swcarpentry/r-novice-gapminder/">Software Carpentry’s R lessons</a></p>
</div>
<div id="introduction-to-rstudio" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-rstudio" class="anchor"></a>Introduction to RStudio</h2>
<p>We’ll be using RStudio: a free, open source R integrated development environment. It provides a built in editor, works on all platforms (including on servers) and provides many advantages such as integration with version control and project management.</p>
<p><strong>Basic layout</strong></p>
<p>When you first open RStudio, you will be greeted by three panels:</p>
<ul>
<li>The interactive R console (entire left)</li>
<li>Environment/History (tabbed in upper right)</li>
<li>Files/Plots/Packages/Help/Viewer (tabbed in lower right)</li>
</ul>
<p>Once you open files, such as R scripts, an editor panel will also open in the top left.</p>
</div>
<div id="workflow-within-rstudio" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-within-rstudio" class="anchor"></a>Workflow within RStudio</h2>
<p>There are two main ways one can work within RStudio.</p>
<ol style="list-style-type: decimal">
<li>Test and play within the interactive R console then copy code into a .R file to run later.
<ul>
<li>This works well when doing small tests and initially starting off.</li>
<li>It quickly becomes laborious</li>
</ul>
</li>
<li>Start writing in an .R file and use RStudio’s command / short cut to push current line, selected lines or modified lines to the interactive R console.
<ul>
<li>This is a great way to start; all your code is saved for later</li>
<li>You will be able to run the file you create from within RStudio or using R’s <code><a href="https://rdrr.io/r/base/source.html">source()</a></code> function.</li>
</ul>
</li>
</ol>
<div id="tip">
<p><i class="far fa-lightbulb"></i> <strong>Tip: Running segments of your code</strong></p>
<p>RStudio offers you great flexibility in running code from within the editor window. There are buttons, menu choices, and keyboard shortcuts. To run the current line, you can 1. click on the <code>Run</code> button just above the editor panel, or 2. select “Run Lines” from the “Code” menu, or 3. hit Ctrl-Enter in Windows or Linux or Command-Enter on OS X. (This shortcut can also be seen by hovering the mouse over the button). To run a block of code, select it and then <code>Run</code>. If you have modified a line of code within a block of code you have just run, there is no need to reselect the section and <code>Run</code>, you can use the next button along, <code>Re-run the previous region</code>. This will run the previous code block including the modifications you have made.</p>
</div>
</div>
<div id="your-rstudio-environment" class="section level2">
<h2 class="hasAnchor">
<a href="#your-rstudio-environment" class="anchor"></a>Your RStudio environment</h2>
<p>Much of your time in R will be spent in the R interactive console. This is where you will run all of your code, and can be a useful environment to try out ideas before adding them to an R script file. This console in RStudio is the same as the one you would get if you just typed in <code>R</code> in your commandline environment.</p>
</div>
<div id="r-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#r-packages" class="anchor"></a>R Packages</h2>
<p>It is possible to add functions to R by writing a package, or by obtaining a package written by someone else. As of this writing, there are over 7,000 packages available on CRAN (the comprehensive R archive network). R and RStudio have functionality for managing packages:</p>
<ul>
<li>You can see what packages are installed by typing <code><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages()</a></code>
</li>
<li>You can install packages by typing <code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("packagename")</a></code>, where <code>packagename</code> is the package name, in quotes.</li>
<li>You can update installed packages by typing <code><a href="https://rdrr.io/r/utils/update.packages.html">update.packages()</a></code>
</li>
<li>You can remove a package with <code><a href="https://rdrr.io/r/utils/remove.packages.html">remove.packages("packagename")</a></code>
</li>
<li>You can make a package available for use with <code><a href="https://rdrr.io/r/base/library.html">library(packagename)</a></code>
</li>
</ul>
<p>For this workshop we will also be using packages from These can all be obtained from Bioconductor.</p>
<p>Open RStudio and run the following commands to install packages from <a href="https://www.bioconductor.org/">Bioconductor</a>. These are installed slightly differently. For example, to install the package <code>limma</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></code></pre></div>
</div>
<div id="rstudio-project-management" class="section level2">
<h2 class="hasAnchor">
<a href="#rstudio-project-management" class="anchor"></a>RStudio project management</h2>
<p>The scientific process is naturally incremental, and many projects start life as random notes, some code, then a manuscript, and eventually everything is a bit mixed together.</p>
<blockquote class="twitter-tweet">
<p>
Managing your projects in a reproducible fashion doesn’t just make your science reproducible, it makes your life easier.
</p>
— Vince Buffalo <a href="https://twitter.com/vsbuffalo/status/323638476153167872">April 15, 2013</a>
</blockquote>
<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script><p>Most people tend to organize their projects like this:</p>
<p><img src="../inst/vignettes/bad_layout.png"><!-- --></p>
<p>There are many reasons why we should <em>ALWAYS</em> avoid this:</p>
<ol style="list-style-type: decimal">
<li>It is really hard to tell which version of your data is the original and which is the modified;</li>
<li>It gets really messy because it mixes files with various extensions together;</li>
<li>It probably takes you a lot of time to actually find things, and relate the correct figures to the exact code that has been used to generate it;</li>
</ol>
<p>A good project layout will ultimately make your life easier:</p>
<ul>
<li>It will help ensure the integrity of your data;</li>
<li>It makes it simpler to share your code with someone else (a lab-mate, collaborator, or supervisor);</li>
<li>It allows you to easily upload your code with your manuscript submission;</li>
<li>It makes it easier to pick the project back up after a break.</li>
</ul>
</div>
<div id="a-possible-solution" class="section level2">
<h2 class="hasAnchor">
<a href="#a-possible-solution" class="anchor"></a>A possible solution</h2>
<p>Fortunately, there are tools and packages which can help you manage your work effectively.</p>
<p>One of the most powerful and useful aspects of RStudio is its project management functionality. We’ll be using this today to create a self-contained, reproducible project.</p>
<div id="challenge">
<p><i class="fas fa-pencil-alt"></i> <strong>Challenge</strong> We’re going to create a new project in RStudio:</p>
<ol style="list-style-type: decimal">
<li>Click the “File” menu button, then “New Project”.</li>
<li>Click “New Directory”.</li>
<li>Click “Empty Project”.</li>
<li>Type in the name of the directory to store your project, e.g. “my_project”, or better yet use the name of this workshop.</li>
<li>Click the “Create Project” button.</li>
</ol>
</div>
<p>Now when we start R in this project directory, or open this project with RStudio, all of our work on this project will be entirely self-contained in this directory.</p>
</div>
<div id="best-practices-for-project-organisation" class="section level2">
<h2 class="hasAnchor">
<a href="#best-practices-for-project-organisation" class="anchor"></a>Best practices for project organisation</h2>
<p>Although there is no “best” way to lay out a project, there are some general principles to adhere to that will make project management easier:</p>
</div>
<div id="treat-data-as-read-only" class="section level2">
<h2 class="hasAnchor">
<a href="#treat-data-as-read-only" class="anchor"></a>Treat data as read only</h2>
<p>This is probably the most important goal of setting up a project. Data is typically time consuming and/or expensive to collect. Working with them interactively (e.g., in Excel) where they can be modified means you are never sure of where the data came from, or how it has been modified since collection. It is therefore a good idea to treat your data as “read-only”.</p>
</div>
<div id="data-cleaning" class="section level2">
<h2 class="hasAnchor">
<a href="#data-cleaning" class="anchor"></a>Data Cleaning</h2>
<p>In many cases your data will be “dirty”: it will need significant preprocessing to get into a format R (or any other programming language) will find useful. This task is sometimes called “data munging”. I find it useful to store these scripts in a separate folder, and create a second “read-only” data folder to hold the “cleaned” data sets.</p>
</div>
<div id="treat-generated-output-as-disposable" class="section level2">
<h2 class="hasAnchor">
<a href="#treat-generated-output-as-disposable" class="anchor"></a>Treat generated output as disposable</h2>
<p>Anything generated by your scripts should be treated as disposable: it should all be able to be regenerated from your scripts.</p>
<p>There are lots of different was to manage this output. I find it useful to have an output folder with different sub-directories for each separate analysis. This makes it easier later, as many of my analyses are exploratory and don’t end up being used in the final project, and some of the analyses get shared between projects.</p>
</div>
<div id="separate-function-definition-and-application" class="section level2">
<h2 class="hasAnchor">
<a href="#separate-function-definition-and-application" class="anchor"></a>Separate function definition and application</h2>
<p>The most effective way I find to work in R, is to play around in the interactive session, then copy commands across to a script file when I’m sure they work and do what I want. You can also save all the commands you’ve entered using the <code>history</code> command, but I don’t find it useful because when I’m typing its 90% trial and error.</p>
<p>When your project is new and shiny, the script file usually contains many lines of directly executed code. As it matures, reusable chunks get pulled into their own functions. It’s a good idea to separate these into separate folders; one to store useful functions that you’ll reuse across analyses and projects, and one to store the analysis scripts.</p>
<div id="tip">
<p><i class="far fa-lightbulb"></i> <strong>Tip: avoiding duplication</strong></p>
<p>You may find yourself using data or analysis scripts across several projects. Typically you want to avoid duplication to save space and avoid having to make updates to code in multiple places.</p>
<p>In this case I find it useful to make “symbolic links”, which are essentially shortcuts to files somewhere else on a filesystem. On Linux and OS X you can use the <code>ln -s</code> command, and on windows you can either create a shortcut or use the <code>mklink</code> command from the windows terminal.</p>
</div>
</div>
<div id="save-the-data-in-the-data-directory" class="section level2">
<h2 class="hasAnchor">
<a href="#save-the-data-in-the-data-directory" class="anchor"></a>Save the data in the data directory</h2>
<p><em>It is recommended to store your data in a <code>/data</code> folder in your project. However in this workshop we will also show you how you can import the data files directly from links.</em> Now we have a good directory structure we will now place/save the data file in the <code>data/</code> directory.</p>
<p>Download the RNAseq data for this workshop.</p>
<ul>
<li>Mouse mammary data (counts): <a href="https://figshare.com/s/1d788fd384d33e913a2a">https://figshare.com/s/1d788fd384d33e913a2a</a>
</li>
<li>Drosophila data (counts): <a href="https://figshare.com/s/e08e71c42f118dbe8be6">https://figshare.com/s/e08e71c42f118dbe8be6</a>
</li>
</ul>
<div id="challenge">
<p><i class="fas fa-pencil-alt"></i> <strong>Challenge</strong></p>
<ol style="list-style-type: decimal">
<li>Create a <code>/data</code> directory. In the bottom right panel select the “Files” tab, then “New Folder”, then type “data” and click “OK”.</li>
<li>Download the RNAseq data using the links above (if you find the internet is slow, you can just download Day 1 for now).</li>
<li>Click “Download all” (this will download a zip file).</li>
<li>Unzip the file (usually double clicking on it will do the trick).</li>
<li>Move all the files inside into the <code>data/</code> folder within your project.</li>
</ol>
</div>
</div>
<div id="reading-data-into-r" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-data-into-r" class="anchor"></a>Reading data into R</h2>
<p>Our data is tab delimited, which is what read.table expect by default so we don’t need to specify <code>delim</code>. However, we do need <code>header = TRUE</code></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read the small_counts.txt data into R</span>

<span class="co"># If you have the file in the data folder you can import from there (run the code line below without the comment character (#))</span>
<span class="co">#small_counts &lt;- read.table("data/small_counts.txt", header = TRUE)</span>

<span class="co"># Here we will import the file directly from its web link</span>
<span class="va">small_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"https://ndownloader.figshare.com/files/6005547?private_link=1d788fd384d33e913a2a"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">small_counts</span><span class="op">)</span></code></pre></div>
<pre><code>##         Sample_1 Sample_2 Sample_3 Sample_4
## Xkr4         438      300       65      237
## Sox17        106      182       82      105
## Mrpl15       309      234      337      300
## Lypla1       652      515      948      935
## Tcea1       1604     1495     1721     1317
## Rgs20          4        2       14        4
## Atp6v1h      769      752     1062      987
## Rb1cc1      1494     1412     1157      967
## Pcmtd1      1344     1242     1374     1593
## Rrs1        1691     1808     2127     1653</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">small_counts</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 10  4</code></pre>
</div>
<div id="working-with-data-frames" class="section level2">
<h2 class="hasAnchor">
<a href="#working-with-data-frames" class="anchor"></a>Working with Data Frames</h2>
<div id="subsetting" class="section level3">
<h3 class="hasAnchor">
<a href="#subsetting" class="anchor"></a>Subsetting</h3>
<p>Let’s say we want the all the counts for Sample_1. There are three main ways we could do this:</p>
<p><code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> notation with the column name</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">small_counts</span><span class="op">$</span><span class="va">Sample_1</span></code></pre></div>
<pre><code>##  [1]  438  106  309  652 1604    4  769 1494 1344 1691</code></pre>
<p><code>[row, column]</code> notation with numeric indices</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">small_counts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>##  [1]  438  106  309  652 1604    4  769 1494 1344 1691</code></pre>
<p><code>[row, column]</code> notation using the column name (in a vector)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">small_counts</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample_1"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>##  [1]  438  106  309  652 1604    4  769 1494 1344 1691</code></pre>
<p>This is handy when you want to grab multiple columns</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">small_counts</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample_1"</span>, <span class="st">"Sample_3"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>##         Sample_1 Sample_3
## Xkr4         438       65
## Sox17        106       82
## Mrpl15       309      337
## Lypla1       652      948
## Tcea1       1604     1721
## Rgs20          4       14
## Atp6v1h      769     1062
## Rb1cc1      1494     1157
## Pcmtd1      1344     1374
## Rrs1        1691     2127</code></pre>
<p>You can even mix numerical and named indices</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">small_counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample_1"</span>, <span class="st">"Sample_3"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>##        Sample_1 Sample_3
## Xkr4        438       65
## Sox17       106       82
## Mrpl15      309      337</code></pre>
</div>
<div id="vectorisation" class="section level3">
<h3 class="hasAnchor">
<a href="#vectorisation" class="anchor"></a>Vectorisation</h3>
<p>R arithmetic operators are vectorised by default. This means they operate on all elements of a vector in parallel.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">small_counts</span><span class="op">$</span><span class="va">Sample_1</span> <span class="op">*</span> <span class="fl">2</span></code></pre></div>
<pre><code>##  [1]  876  212  618 1304 3208    8 1538 2988 2688 3382</code></pre>
<p>So are many R functions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">small_counts</span><span class="op">)</span></code></pre></div>
<pre><code>##         Sample_1  Sample_2 Sample_3 Sample_4
## Xkr4    6.082219 5.7037825 4.174387 5.468060
## Sox17   4.663439 5.2040067 4.406719 4.653960
## Mrpl15  5.733341 5.4553211 5.820083 5.703782
## Lypla1  6.480045 6.2441669 6.854355 6.840547
## Tcea1   7.380256 7.3098815 7.450661 7.183112
## Rgs20   1.386294 0.6931472 2.639057 1.386294
## Atp6v1h 6.645091 6.6227363 6.967909 6.894670
## Rb1cc1  7.309212 7.2527624 7.053586 6.874198
## Pcmtd1  7.203406 7.1244783 7.225481 7.373374
## Rrs1    7.433075 7.4999765 7.662468 7.410347</code></pre>
<p>Let’s say we want to calculate the sum of counts for each sample. We can do this for each sample by subsetting and then using the <code>sum</code> function.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">small_counts</span><span class="op">$</span><span class="va">Sample_1</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 8411</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">small_counts</span><span class="op">$</span><span class="va">Sample_2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 7942</code></pre>
<p>If you had a lot of samples, you can imagine this would get tedious. So instead, we can use an <code>apply</code> statement to apply the <code>sum</code> function to every column at once. In R, this is usually quicker than using a loop. Note the <code>MARGIN</code> argument tells <code>apply</code> which direction you want to go in. MARGIN = 1 will apply <code>sum</code> to each row, while MARGIN = 2 will apply <code>sum</code> to each column.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Sum the counts for each sample</span>
<span class="va">sample_sums</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">small_counts</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, <span class="va">sum</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">sample_sums</span><span class="op">)</span></code></pre></div>
<pre><code>## Sample_1 Sample_2 Sample_3 Sample_4 
##     8411     7942     8887     8098</code></pre>
</div>
</div>
<div id="data-types" class="section level2">
<h2 class="hasAnchor">
<a href="#data-types" class="anchor"></a>Data types</h2>
<p>There are 5 main types: doubles, integers, complex, logical and character.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="fl">3.14</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "double"</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="fl">1L</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "integer"</code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="fl">1i</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "complex"</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "logical"</code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="st">'banana'</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
<p>Note the <code>L</code> suffix to insist that a number is an integer. No matter how complicated our analyses become, all data in R is interpreted as one of these basic data types.</p>
<p>Let’s read in some data that contains things of more than one type. This is actually a sample of a differential expression results table that you will generate later in the workshop.</p>
<p>We add <code>stringsAsFactors = TRUE</code> to show you what it does as this was what <code>read.table</code> did by default until R version 4.0, released April 2020.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># import from data folder</span>
<span class="co">#ResultsTable_small &lt;- read.table("data/ResultsTable_small.txt", stringsAsFactors = TRUE, header=TRUE)</span>

<span class="co"># import from web link</span>
<span class="va">ResultsTable_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"https://ndownloader.figshare.com/files/6005550?private_link=1d788fd384d33e913a2a"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">TRUE</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>


<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">)</span></code></pre></div>
<pre><code>##    ENTREZID        SYMBOL     logFC     AveExpr         t      P.Value
## 1     24117          Wif1  1.819943  2.97554452  20.10780 1.063770e-10
## 2    381290        Atp2b4 -2.143885  3.94406593 -19.07495 1.982934e-10
## 3     78896 1500015O10Rik  2.807548  3.03651950  18.54773 2.758828e-10
## 4    226101          Myof -2.329744  6.22352456 -18.26861 3.297667e-10
## 5     16012        Igfbp6 -2.896115  1.97844876 -18.21525 3.413066e-10
## 6    231830       Micall2  2.253400  4.76059697  18.02627 3.858161e-10
## 7     16669         Krt19 -2.312721  8.74189184 -17.07937 7.264548e-10
## 8     55987         Cpxm2 -1.515469  2.83451194 -16.64333 9.829870e-10
## 9    231991         Creb5 -2.598105  4.27592952 -16.53634 1.059885e-09
## 10    14620          Gjb3  3.600094  3.52528051  16.46627 1.113755e-09
## 11   211577        Mrgprf -5.146268 -0.93683349 -16.36573 1.196263e-09
## 12    20317      Serpinf1  1.710380  3.38883490  15.77280 1.838727e-09
## 13    74747         Ddit4  2.180370  6.86479110  15.70145 1.938279e-09
## 14   270150       Ccdc153 -3.211148 -1.34083882 -15.50126 2.249931e-09
## 15    11636           Ak1  2.766745  4.30347462  15.27694 2.664640e-09
## 16    20482          Skil  1.887561  8.49892507  14.65488 4.311334e-09
## 17   194126        Mtmr11 -1.708973  2.50804119 -14.48746 4.922928e-09
## 18    21953         Tnni2 -5.827889  0.30207159 -14.40327 5.265278e-09
## 19    12992       Csn1s2b -6.070143  3.56295004 -14.16565 6.377604e-09
## 20    17131         Smad7  1.972771  6.71751902  14.14348 6.493642e-09
## 21    20564         Slit3 -1.331437  3.44179493 -13.88522 8.026279e-09
## 22    16878           Lif  3.738933  6.68203417  13.73344 9.105708e-09
## 23   170761         Pdzd3 -3.313648 -0.06019306 -13.62372 9.982985e-09
## 24    73847       Fam110a  1.474671  6.84086068  13.62251 9.993185e-09
## 25    12977          Csf1  2.835624  7.47759094  13.41902 1.187300e-08
## 26    12654         Chil1  2.342914  5.57645724  13.21976 1.408760e-08
## 27   545370         Hmcn1 -1.567424  3.10302591 -13.19053 1.444832e-08
## 28    60599     Trp53inp1 -1.258671  6.11839605 -13.16241 1.480464e-08
## 29   217166         Nr1d1  2.278879  6.26087761  13.12885 1.524242e-08
## 30    67111          Naaa -2.538596  3.29074575 -13.04083 1.645823e-08
## 31   232016       Ccdc129 -2.597044  5.00471484 -13.02266 1.672195e-08
## 32    76123         Gpsm2 -2.171896  4.99093472 -12.76344 2.102397e-08
## 33   329739       Fam102b -1.520291  4.18813047 -12.75357 2.120968e-08
## 34    79362       Bhlhe41  1.788601  6.18368494  12.70504 2.214908e-08
## 35    50786        Hs6st2  1.751340  0.53953600  12.43097 2.836919e-08
## 36    18019        Nfatc2 -2.011808  5.79499693 -12.27561 3.271067e-08
## 37    67971         Tppp3 -2.653398  4.90816305 -12.22845 3.416616e-08
## 38   235542       Ppp2r3a -1.109156  6.50105941 -12.22041 3.442139e-08
## 39    74134        Cyp2s1 -2.071469  1.40704575 -12.20154 3.502805e-08
## 40    20677          Sox4  1.522405  7.46932835  12.13548 3.724389e-08
##       adj.P.Val
## 1  1.016240e-06
## 2  1.016240e-06
## 3  1.016240e-06
## 4  1.016240e-06
## 5  1.016240e-06
## 6  1.016240e-06
## 7  1.640127e-06
## 8  1.718703e-06
## 9  1.718703e-06
## 10 1.718703e-06
## 11 1.718703e-06
## 12 2.356351e-06
## 13 2.356351e-06
## 14 2.539851e-06
## 15 2.807465e-06
## 16 4.258521e-06
## 17 4.576586e-06
## 18 4.622914e-06
## 19 5.131276e-06
## 20 5.131276e-06
## 21 6.040348e-06
## 22 6.541210e-06
## 23 6.580512e-06
## 24 6.580512e-06
## 25 7.505634e-06
## 26 8.306595e-06
## 27 8.306595e-06
## 28 8.306595e-06
## 29 8.306595e-06
## 30 8.524957e-06
## 31 8.524957e-06
## 32 1.015751e-05
## 33 1.015751e-05
## 34 1.029541e-05
## 35 1.280991e-05
## 36 1.419445e-05
## 37 1.419445e-05
## 38 1.419445e-05
## 39 1.419445e-05
## 40 1.462109e-05</code></pre>
<p>We can use <code>str</code> to reveal the structure of our <code>ResultsTable_small</code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">)</span></code></pre></div>
<pre><code>## 'data.frame':    40 obs. of  7 variables:
##  $ ENTREZID : int  24117 381290 78896 226101 16012 231830 16669 55987 231991 14620 ...
##  $ SYMBOL   : Factor w/ 40 levels "1500015O10Rik",..: 40 3 1 26 20 23 21 8 9 16 ...
##  $ logFC    : num  1.82 -2.14 2.81 -2.33 -2.9 ...
##  $ AveExpr  : num  2.98 3.94 3.04 6.22 1.98 ...
##  $ t        : num  20.1 -19.1 18.5 -18.3 -18.2 ...
##  $ P.Value  : num  1.06e-10 1.98e-10 2.76e-10 3.30e-10 3.41e-10 ...
##  $ adj.P.Val: num  1.02e-06 1.02e-06 1.02e-06 1.02e-06 1.02e-06 ...</code></pre>
<p>We can see that it’s a data frame, and <code>str</code> also tells us the type of each column. Everything in a single column must be of the same type, while each column can be a different type. This is because data frames are actually lists of vectors, and vectors must contain data of only one type.</p>
<p>When you try to mix types in a vector, strange things can happen.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">my_vector</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"hello"</span>, <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">my_vector</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "1"     "hello" "TRUE"</code></pre>
<p>Notice that everything has been converted to a character. You can check this with <code>typeof</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">my_vector</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
<p>This is called type coercion. When you try to create a vector containing a mixture of types R will automatically (and silently) convert everything to the same type. Generally it will chose a type that means you don’t lose any information.</p>
<p>The coercion rules go: <code>logical</code> -&gt; <code>integer</code> -&gt; <code>numeric</code> -&gt; <code>complex</code> -&gt; <code>character</code>.</p>
<p>Note, the same thing will happen if you try to read in data from a file that has a mixture of types within a single column.</p>
<div id="factors" class="section level3">
<h3 class="hasAnchor">
<a href="#factors" class="anchor"></a>Factors</h3>
<p>You may have noticed that <code>ResultsTable_small</code> contained a column annotated as <code>Factor</code>.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">)</span></code></pre></div>
<pre><code>## 'data.frame':    40 obs. of  7 variables:
##  $ ENTREZID : int  24117 381290 78896 226101 16012 231830 16669 55987 231991 14620 ...
##  $ SYMBOL   : Factor w/ 40 levels "1500015O10Rik",..: 40 3 1 26 20 23 21 8 9 16 ...
##  $ logFC    : num  1.82 -2.14 2.81 -2.33 -2.9 ...
##  $ AveExpr  : num  2.98 3.94 3.04 6.22 1.98 ...
##  $ t        : num  20.1 -19.1 18.5 -18.3 -18.2 ...
##  $ P.Value  : num  1.06e-10 1.98e-10 2.76e-10 3.30e-10 3.41e-10 ...
##  $ adj.P.Val: num  1.02e-06 1.02e-06 1.02e-06 1.02e-06 1.02e-06 ...</code></pre>
<p>Factors usually look like character data, but are typically used to represent categorical information.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span></code></pre></div>
<pre><code>##  Factor w/ 40 levels "1500015O10Rik",..: 40 3 1 26 20 23 21 8 9 16 ...</code></pre>
<p>Instead of printing out the strings we gave it, we got a bunch of numbers instead. R has replaced our human-readable categories with numbered indices under the hood:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "integer"</code></pre>
<p>In modelling functions, it’s important to know what the baseline levels are. This is assumed to be the first factor, but by default factors are labelled in alphabetical order. You can change this by specifying the levels:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mydata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"case"</span>, <span class="st">"control"</span>, <span class="st">"control"</span>, <span class="st">"case"</span><span class="op">)</span>
<span class="va">factor_ordering_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">mydata</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"case"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">factor_ordering_example</span><span class="op">)</span></code></pre></div>
<pre><code>##  Factor w/ 2 levels "control","case": 2 1 1 2</code></pre>
<p>In this case, we’ve explicitly told R that “control” should represented by 1, and “case” by 2. This designation can be very important for interpreting the results of statistical models!</p>
<p>If you prefer not to work with factors, you can simply tell the <code>read.table</code> function to keep all strings as characters. This is now the default for <code>read.table</code> since R version 4.0, released April 2020 so you don’t need to specify <code>stringsAsFactors = FALSE</code> if you’re using R &gt;= 4.0.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># from data folder</span>
<span class="co">#ResultsTable_small &lt;- read.table("data/ResultsTable_small.txt", stringsAsFactors = FALSE, header=TRUE)</span>

<span class="co"># from web link</span>
<span class="va">ResultsTable_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"https://ndownloader.figshare.com/files/6005550?private_link=1d788fd384d33e913a2a"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="sorting" class="section level3">
<h3 class="hasAnchor">
<a href="#sorting" class="anchor"></a>Sorting</h3>
<p>If you just want to sort a vector from smallest to largest, you can use the <code>sort</code> function:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] -6.070143 -5.827889 -5.146268 -3.313648 -3.211148 -2.896115 -2.653398
##  [8] -2.598105 -2.597044 -2.538596 -2.329744 -2.312721 -2.171896 -2.143885
## [15] -2.071469 -2.011808 -1.708973 -1.567424 -1.520291 -1.515469 -1.331437
## [22] -1.258671 -1.109156  1.474671  1.522405  1.710380  1.751340  1.788601
## [29]  1.819943  1.887561  1.972771  2.180370  2.253400  2.278879  2.342914
## [36]  2.766745  2.807548  2.835624  3.600094  3.738933</code></pre>
<p>To reverse the sort order so it goes from largest to smallest:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1]  3.738933  3.600094  2.835624  2.807548  2.766745  2.342914  2.278879
##  [8]  2.253400  2.180370  1.972771  1.887561  1.819943  1.788601  1.751340
## [15]  1.710380  1.522405  1.474671 -1.109156 -1.258671 -1.331437 -1.515469
## [22] -1.520291 -1.567424 -1.708973 -2.011808 -2.071469 -2.143885 -2.171896
## [29] -2.312721 -2.329744 -2.538596 -2.597044 -2.598105 -2.653398 -2.896115
## [36] -3.211148 -3.313648 -5.146268 -5.827889 -6.070143</code></pre>
<p>And this works with characters:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "1500015O10Rik" "Ak1"           "Atp2b4"        "Bhlhe41"      
##  [5] "Ccdc129"       "Ccdc153"       "Chil1"         "Cpxm2"        
##  [9] "Creb5"         "Csf1"          "Csn1s2b"       "Cyp2s1"       
## [13] "Ddit4"         "Fam102b"       "Fam110a"       "Gjb3"         
## [17] "Gpsm2"         "Hmcn1"         "Hs6st2"        "Igfbp6"       
## [21] "Krt19"         "Lif"           "Micall2"       "Mrgprf"       
## [25] "Mtmr11"        "Myof"          "Naaa"          "Nfatc2"       
## [29] "Nr1d1"         "Pdzd3"         "Ppp2r3a"       "Serpinf1"     
## [33] "Skil"          "Slit3"         "Smad7"         "Sox4"         
## [37] "Tnni2"         "Tppp3"         "Trp53inp1"     "Wif1"</code></pre>
<p>If you want to sort the entire data frame by the values in a single column, you can use the <code>order</code> function. This gives you the indexes for if the data was sorted. For example:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] 19 18 11 23 14  5 37  9 31 30  4  7 32  2 39 36 17 27 33  8 21 28 38 24 40
## [26] 12 35 34  1 16 20 13  6 29 26 15  3 25 10 22</code></pre>
<p>You then use these indexes to sort the data:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>##  [1] -6.070143 -5.827889 -5.146268 -3.313648 -3.211148 -2.896115 -2.653398
##  [8] -2.598105 -2.597044 -2.538596 -2.329744 -2.312721 -2.171896 -2.143885
## [15] -2.071469 -2.011808 -1.708973 -1.567424 -1.520291 -1.515469 -1.331437
## [22] -1.258671 -1.109156  1.474671  1.522405  1.710380  1.751340  1.788601
## [29]  1.819943  1.887561  1.972771  2.180370  2.253400  2.278879  2.342914
## [36]  2.766745  2.807548  2.835624  3.600094  3.738933</code></pre>
<p>You can then use these indices to sort the rows:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span>, <span class="op">]</span></code></pre></div>
<pre><code>##    ENTREZID        SYMBOL     logFC     AveExpr         t      P.Value
## 19    12992       Csn1s2b -6.070143  3.56295004 -14.16565 6.377604e-09
## 18    21953         Tnni2 -5.827889  0.30207159 -14.40327 5.265278e-09
## 11   211577        Mrgprf -5.146268 -0.93683349 -16.36573 1.196263e-09
## 23   170761         Pdzd3 -3.313648 -0.06019306 -13.62372 9.982985e-09
## 14   270150       Ccdc153 -3.211148 -1.34083882 -15.50126 2.249931e-09
## 5     16012        Igfbp6 -2.896115  1.97844876 -18.21525 3.413066e-10
## 37    67971         Tppp3 -2.653398  4.90816305 -12.22845 3.416616e-08
## 9    231991         Creb5 -2.598105  4.27592952 -16.53634 1.059885e-09
## 31   232016       Ccdc129 -2.597044  5.00471484 -13.02266 1.672195e-08
## 30    67111          Naaa -2.538596  3.29074575 -13.04083 1.645823e-08
## 4    226101          Myof -2.329744  6.22352456 -18.26861 3.297667e-10
## 7     16669         Krt19 -2.312721  8.74189184 -17.07937 7.264548e-10
## 32    76123         Gpsm2 -2.171896  4.99093472 -12.76344 2.102397e-08
## 2    381290        Atp2b4 -2.143885  3.94406593 -19.07495 1.982934e-10
## 39    74134        Cyp2s1 -2.071469  1.40704575 -12.20154 3.502805e-08
## 36    18019        Nfatc2 -2.011808  5.79499693 -12.27561 3.271067e-08
## 17   194126        Mtmr11 -1.708973  2.50804119 -14.48746 4.922928e-09
## 27   545370         Hmcn1 -1.567424  3.10302591 -13.19053 1.444832e-08
## 33   329739       Fam102b -1.520291  4.18813047 -12.75357 2.120968e-08
## 8     55987         Cpxm2 -1.515469  2.83451194 -16.64333 9.829870e-10
## 21    20564         Slit3 -1.331437  3.44179493 -13.88522 8.026279e-09
## 28    60599     Trp53inp1 -1.258671  6.11839605 -13.16241 1.480464e-08
## 38   235542       Ppp2r3a -1.109156  6.50105941 -12.22041 3.442139e-08
## 24    73847       Fam110a  1.474671  6.84086068  13.62251 9.993185e-09
## 40    20677          Sox4  1.522405  7.46932835  12.13548 3.724389e-08
## 12    20317      Serpinf1  1.710380  3.38883490  15.77280 1.838727e-09
## 35    50786        Hs6st2  1.751340  0.53953600  12.43097 2.836919e-08
## 34    79362       Bhlhe41  1.788601  6.18368494  12.70504 2.214908e-08
## 1     24117          Wif1  1.819943  2.97554452  20.10780 1.063770e-10
## 16    20482          Skil  1.887561  8.49892507  14.65488 4.311334e-09
## 20    17131         Smad7  1.972771  6.71751902  14.14348 6.493642e-09
## 13    74747         Ddit4  2.180370  6.86479110  15.70145 1.938279e-09
## 6    231830       Micall2  2.253400  4.76059697  18.02627 3.858161e-10
## 29   217166         Nr1d1  2.278879  6.26087761  13.12885 1.524242e-08
## 26    12654         Chil1  2.342914  5.57645724  13.21976 1.408760e-08
## 15    11636           Ak1  2.766745  4.30347462  15.27694 2.664640e-09
## 3     78896 1500015O10Rik  2.807548  3.03651950  18.54773 2.758828e-10
## 25    12977          Csf1  2.835624  7.47759094  13.41902 1.187300e-08
## 10    14620          Gjb3  3.600094  3.52528051  16.46627 1.113755e-09
## 22    16878           Lif  3.738933  6.68203417  13.73344 9.105708e-09
##       adj.P.Val
## 19 5.131276e-06
## 18 4.622914e-06
## 11 1.718703e-06
## 23 6.580512e-06
## 14 2.539851e-06
## 5  1.016240e-06
## 37 1.419445e-05
## 9  1.718703e-06
## 31 8.524957e-06
## 30 8.524957e-06
## 4  1.016240e-06
## 7  1.640127e-06
## 32 1.015751e-05
## 2  1.016240e-06
## 39 1.419445e-05
## 36 1.419445e-05
## 17 4.576586e-06
## 27 8.306595e-06
## 33 1.015751e-05
## 8  1.718703e-06
## 21 6.040348e-06
## 28 8.306595e-06
## 38 1.419445e-05
## 24 6.580512e-06
## 40 1.462109e-05
## 12 2.356351e-06
## 35 1.280991e-05
## 34 1.029541e-05
## 1  1.016240e-06
## 16 4.258521e-06
## 20 5.131276e-06
## 13 2.356351e-06
## 6  1.016240e-06
## 29 8.306595e-06
## 26 8.306595e-06
## 15 2.807465e-06
## 3  1.016240e-06
## 25 7.505634e-06
## 10 1.718703e-06
## 22 6.541210e-06</code></pre>
</div>
<div id="subsetting-using-logical-statements" class="section level3">
<h3 class="hasAnchor">
<a href="#subsetting-using-logical-statements" class="anchor"></a>Subsetting using logical statements</h3>
<p>Let’s say we only want to look at the rows where the log fold change is greater than 3. We can ask if each <code>logFC</code> is greater than 3.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">3</span></code></pre></div>
<pre><code>##  [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE
## [13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE
## [25] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [37] FALSE FALSE FALSE FALSE</code></pre>
<p>This gives us a vector of TRUE/FALSE for each value stating if it is greater than 3. We can then use this to subset the data.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span><span class="op">[</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 3.600094 3.738933</code></pre>
<p>And to subset the rows of the data frame.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">[</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">3</span>, <span class="op">]</span></code></pre></div>
<pre><code>##    ENTREZID SYMBOL    logFC  AveExpr        t      P.Value    adj.P.Val
## 10    14620   Gjb3 3.600094 3.525281 16.46627 1.113755e-09 1.718703e-06
## 22    16878    Lif 3.738933 6.682034 13.73344 9.105708e-09 6.541210e-06</code></pre>
<p>In fact, we probably care about things that are -3 as well, so let’s include those too:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">[</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">3</span> <span class="op">|</span> <span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">3</span>, <span class="op">]</span></code></pre></div>
<pre><code>##    ENTREZID  SYMBOL     logFC     AveExpr         t      P.Value    adj.P.Val
## 10    14620    Gjb3  3.600094  3.52528051  16.46627 1.113755e-09 1.718703e-06
## 11   211577  Mrgprf -5.146268 -0.93683349 -16.36573 1.196263e-09 1.718703e-06
## 14   270150 Ccdc153 -3.211148 -1.34083882 -15.50126 2.249931e-09 2.539851e-06
## 18    21953   Tnni2 -5.827889  0.30207159 -14.40327 5.265278e-09 4.622914e-06
## 19    12992 Csn1s2b -6.070143  3.56295004 -14.16565 6.377604e-09 5.131276e-06
## 22    16878     Lif  3.738933  6.68203417  13.73344 9.105708e-09 6.541210e-06
## 23   170761   Pdzd3 -3.313648 -0.06019306 -13.62372 9.982985e-09 6.580512e-06</code></pre>
<p>Or we can simplify this using the <code>abs</code> function to find the absolute value of each <code>logFC</code> before testing them.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3</span>, <span class="op">]</span></code></pre></div>
<pre><code>##    ENTREZID  SYMBOL     logFC     AveExpr         t      P.Value    adj.P.Val
## 10    14620    Gjb3  3.600094  3.52528051  16.46627 1.113755e-09 1.718703e-06
## 11   211577  Mrgprf -5.146268 -0.93683349 -16.36573 1.196263e-09 1.718703e-06
## 14   270150 Ccdc153 -3.211148 -1.34083882 -15.50126 2.249931e-09 2.539851e-06
## 18    21953   Tnni2 -5.827889  0.30207159 -14.40327 5.265278e-09 4.622914e-06
## 19    12992 Csn1s2b -6.070143  3.56295004 -14.16565 6.377604e-09 5.131276e-06
## 22    16878     Lif  3.738933  6.68203417  13.73344 9.105708e-09 6.541210e-06
## 23   170761   Pdzd3 -3.313648 -0.06019306 -13.62372 9.982985e-09 6.580512e-06</code></pre>
<div id="tip">
<p><i class="far fa-lightbulb"></i> <strong>Tip: Logical Operators in R</strong></p>
<table class="table">
<thead><tr class="header">
<th>Operator</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>&lt;</td>
<td>less than</td>
</tr>
<tr class="even">
<td>&lt;=</td>
<td>less than or equal to</td>
</tr>
<tr class="odd">
<td>&gt;</td>
<td>greater than</td>
</tr>
<tr class="even">
<td>&gt;=</td>
<td>greater than or equal to</td>
</tr>
<tr class="odd">
<td>==</td>
<td>exactly equal to</td>
</tr>
<tr class="even">
<td>!=</td>
<td>not equal to</td>
</tr>
<tr class="odd">
<td>!x</td>
<td>Not x</td>
</tr>
<tr class="even">
<td>x | y</td>
<td>x OR y</td>
</tr>
<tr class="odd">
<td>x &amp; y</td>
<td>x AND y</td>
</tr>
<tr class="even">
<td>isTRUE(x)</td>
<td>test if X is TRUE</td>
</tr>
</tbody>
</table>
<p>Source: <a href="http://www.statmethods.net/management/operators.html">Quick-R</a></p>
</div>
<p>We might also want to reduce our results table down to a set of genes that we are interested in. We can use similar logical subsetting logic to do this.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">my_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Smad7"</span>, <span class="st">"Wif1"</span>, <span class="st">"Fam102b"</span>, <span class="st">"Tppp3"</span><span class="op">)</span></code></pre></div>
<p>For each gene in ResultsTable_small$SYMBOL, does it appear in my_genes?</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op">%in%</span> <span class="va">my_genes</span></code></pre></div>
<pre><code>##  [1]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE
## [25] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE
## [37]  TRUE FALSE FALSE FALSE</code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">[</span><span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op">%in%</span> <span class="va">my_genes</span>, <span class="op">]</span></code></pre></div>
<pre><code>##    ENTREZID  SYMBOL     logFC  AveExpr         t      P.Value    adj.P.Val
## 1     24117    Wif1  1.819943 2.975545  20.10780 1.063770e-10 1.016240e-06
## 20    17131   Smad7  1.972771 6.717519  14.14348 6.493642e-09 5.131276e-06
## 33   329739 Fam102b -1.520291 4.188130 -12.75357 2.120968e-08 1.015751e-05
## 37    67971   Tppp3 -2.653398 4.908163 -12.22845 3.416616e-08 1.419445e-05</code></pre>
<p>You can use <code>match</code> to get the same subset, but this time make sure they are in the same order as in <code>my_genes</code>. This can be useful if you are trying to merge together data from two sources.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">my_genes</span>, <span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 20  1 33 37</code></pre>
<p>This returns the index for where each of the genes in <code>my_list appears</code> in <code>ResultsTable_small$SYMBOL</code>. We can then use this to subset the columns from <code>ResultsTable_small</code>.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ResultsTable_small</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">my_genes</span>, <span class="va">ResultsTable_small</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span>, <span class="op">]</span></code></pre></div>
<pre><code>##    ENTREZID  SYMBOL     logFC  AveExpr         t      P.Value    adj.P.Val
## 20    17131   Smad7  1.972771 6.717519  14.14348 6.493642e-09 5.131276e-06
## 1     24117    Wif1  1.819943 2.975545  20.10780 1.063770e-10 1.016240e-06
## 33   329739 Fam102b -1.520291 4.188130 -12.75357 2.120968e-08 1.015751e-05
## 37    67971   Tppp3 -2.653398 4.908163 -12.22845 3.416616e-08 1.419445e-05</code></pre>
<p>You can see they are in the same order as <code>my_genes</code>!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
