<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNA-seq analysis in R • RNAseqRworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="RNA-seq analysis in R">
<meta property="og:description" content="RNAseqRworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAseqRworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/00-r-rstudio-intro.html">R for RNA-seq</a>
</li>
<li>
  <a href="../articles/06-rnaseq-day1.html">Differential expression analysis</a>
</li>
<li>
  <a href="../articles/08-applying-rnaseq.html">Applying RNA-seq</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://mblue9.github.io/RNAseq-R_testing">
    <span class="fab fa-home"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/mblue9/RNAseq-R_testing/">
    <span class="fab fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="08-applying-rnaseq_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNA-seq analysis in R</h1>
            <h3 class="subtitle">Analysis of Pasilla Knock-down Experiment in Drosophila</h3>
                        <h4 class="author">Belinda Phipson and Jovana Maksimovic</h4>
            
            <h4 class="date">27 April 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mblue9/RNAseq-R/blob/master/vignettes/08-applying-rnaseq.Rmd"><code>vignettes/08-applying-rnaseq.Rmd</code></a></small>
      <div class="hidden name"><code>08-applying-rnaseq.Rmd</code></div>

    </div>

    
    
<div id="data-files-needed" class="section level2">
<h2 class="hasAnchor">
<a href="#data-files-needed" class="anchor"></a>Data files needed</h2>
<ul>
<li><a href="https://ndownloader.figshare.com/files/5999835?private_link=e08e71c42f118dbe8be6">counts_Drosophila.txt</a></li>
<li><a href="https://ndownloader.figshare.com/files/5999838?private_link=e08e71c42f118dbe8be6">SampleInfo_Drosophila.txt</a></li>
</ul>
<p>Available from <a href="https://figshare.com/s/e08e71c42f118dbe8be6">Figshare</a>.</p>
</div>
<div id="libraries-needed" class="section level2">
<h2 class="hasAnchor">
<a href="#libraries-needed" class="anchor"></a>Libraries needed</h2>
<ul>
<li>limma</li>
<li>edgeR</li>
<li>org.Dm.eg.db (annotation for Drosophila)</li>
<li>gplots</li>
<li>RColorBrewer</li>
</ul>
<p><strong>If you have your own dataset, make sure you have a counts matrix and the sample information. You can look for the annotation package for the species you’re studying on the <a href="http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData">Bioconductor website</a>.</strong> (Search for “org.” to see the organism packages)</p>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>The RNA-Seq data we will be analysing today comes from this published paper:</p>
<p>Brooks, A.N., Yang, L., Duff, M.O., Hansen, K.D., Park, J.W., Dudoit, S., Brenner, S.E. and Graveley, B.R. (2011) Conservation of an rna regulatory map between drosophila and mammals. Genome Research, 21(2), 193-202.<br><a href="http://www.ncbi.nlm.nih.gov/pubmed/20921232" class="uri">http://www.ncbi.nlm.nih.gov/pubmed/20921232</a></p>
<p>This is a publicly available dataset, deposited in the Short Read Archive. The RNA-sequence data are available from GEO under accession nos. GSM461176-GSM461181. The authors combined RNAi and RNASeq to identify exons regulated by Pasilla, the Drosophila melanogaster ortholog of mammalian NOVA1 and NOVA2. They showed that the RNA regulatory map of Pasilla and NOVA1/2 is highly conserved between insects and mammals. NOVA1 and NOVA2 are best known for being involved in alternative splicing. Cells from S2-DRSC, which is an embryonic cell line, were cultured and subjected to a treatment in order to knock-down Pasilla. The four untreated and three treated RNAi samples were used in the analysis. The treated samples had Pasilla knocked down by approximately 60% compared to the untreated samples. Some of the samples had undergone paired end sequencing while other samples were sequenced from one end only.</p>
<p>The reads were aligned to the Drosophila reference genome, downloaded from Ensembl, using the tophat aligner. The reads were summarised at the gene-level using htseq-count, a function from the tool <a href="http://wwwhuber.embl.de/users/anders/HTSeq/doc/overview.html">HTSeq</a>.</p>
<p>For the purpose of today’s workshop, we will be analysing the gene level counts.</p>
</div>
<div id="tasks" class="section level2">
<h2 class="hasAnchor">
<a href="#tasks" class="anchor"></a>Tasks</h2>
<ul>
<li><p>Importing the data into R</p></li>
<li><p>Import in sample information file (if you have your own dataset, make a simple tab delimited file with sample information in Excel to read into R)</p></li>
<li><p>Filtering out lowly expressed genes</p></li>
<li>
<p>Quality control</p>
<ul>
<li>Check library sizes</li>
<li>Check boxplots of log2 cpm</li>
<li>Check MDSplots</li>
</ul>
</li>
<li><p>Hierarchical clustering with heatmap.2</p></li>
<li><p>Normalisation</p></li>
<li>
<p>Testing for differential expression</p>
<ul>
<li>Set up appropriate design matrix</li>
<li>voom transform the data</li>
<li>Fit linear model and get DE genes</li>
<li>Add annotation (different from day 1 - we will show you how)</li>
</ul>
</li>
<li>
<p>Plots after DE</p>
<ul>
<li>Check top gene that is DE between Treated and Untreated</li>
<li>Check expression of <em>pasilla</em> (ensemble gene id: FBgn0261552)</li>
<li>Mean-difference plot (plotMD)</li>
<li>Volcano plot</li>
</ul>
</li>
<li><p>Testing relative to a threshold</p></li>
<li><p>Gene set testing with goana</p></li>
<li><p>Try some interactive plots from Glimma and save the topTable output in a spreadsheet.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
