<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNA-seq analysis in R • RNAseqRworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="RNA-seq analysis in R">
<meta property="og:description" content="RNAseqRworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAseqRworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/00-r-rstudio-intro.html">R for RNA-seq</a>
</li>
<li>
  <a href="../articles/06-rnaseq-day1.html">Differential expression analysis</a>
</li>
<li>
  <a href="../articles/08-applying-rnaseq.html">Applying RNA-seq</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Slides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/slides-01-rnaseq-basics.Rmd">RNA-seq basics</a>
    </li>
    <li>
      <a href="../articles/slides-02-rnaseq-filtering-qc.Rmd">Filtering and QC</a>
    </li>
    <li>
      <a href="../articles/slides-03-differential-expression-analysis.Rmd">Differential expression analysis</a>
    </li>
    <li>
      <a href="../articles/slides-04-gene-set-testing.Rmd">Gene set testing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://mblue9.github.io/RNAseq-R_testing">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/mblue9/RNAseq-R_testing/">
    <span class="fab fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="09-applying-rnaseq-solutions_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNA-seq analysis in R</h1>
            <h3 class="subtitle">Solutions: Analysis of Pasilla Knock-down Experiment in Drosophila</h3>
                        <h4 class="author">Belinda Phipson and Jovana Maksimovic</h4>
            
            <h4 class="date">27 April 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mblue9/RNAseq-R/blob/master/vignettes/09-applying-rnaseq-solutions.Rmd"><code>vignettes/09-applying-rnaseq-solutions.Rmd</code></a></small>
      <div class="hidden name"><code>09-applying-rnaseq-solutions.Rmd</code></div>

    </div>

    
    
<div id="data-files-needed" class="section level2">
<h2 class="hasAnchor">
<a href="#data-files-needed" class="anchor"></a>Data files needed</h2>
<ul>
<li><a href="https://ndownloader.figshare.com/files/5999835?private_link=e08e71c42f118dbe8be6">counts_Drosophila.txt</a></li>
<li><a href="https://ndownloader.figshare.com/files/5999838?private_link=e08e71c42f118dbe8be6">SampleInfo_Drosophila.txt</a></li>
</ul>
<p>Available from <a href="https://figshare.com/s/e08e71c42f118dbe8be6">Figshare</a>.</p>
</div>
<div id="libraries-needed" class="section level2">
<h2 class="hasAnchor">
<a href="#libraries-needed" class="anchor"></a>Libraries needed</h2>
<ul>
<li>limma</li>
<li>edgeR</li>
<li>org.Dm.eg.db</li>
<li>gplots</li>
<li>RColorBrewer</li>
</ul>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>The RNA-Seq data we will be analysing today come from this published paper:</p>
<p>Brooks, A.N., Yang, L., Duff, M.O., Hansen, K.D., Park, J.W., Dudoit, S., Brenner, S.E. and Graveley, B.R. (2011) Conservation of an rna regulatory map between drosophila and mammals. Genome Research, 21(2), 193-202.<br><a href="http://www.ncbi.nlm.nih.gov/pubmed/20921232" class="uri">http://www.ncbi.nlm.nih.gov/pubmed/20921232</a></p>
<p>This is a publicly available dataset, deposited in the Short Read Archive. The RNA-sequence data are available from GEO under accession nos. GSM461176-GSM461181. The authors combined RNAi and RNASeq to identify exons regulated by Pasilla, the Drosophila melanogaster ortholog of mammalian NOVA1 and NOVA2. They showed that the RNA regulatory map of Pasilla and NOVA1/2 is highly conserved between insects and mammals. NOVA1 and NOVA2 are best known for being involved in alternative splicing. Cells from S2-DRSC, which is an embryonic cell line, were cultured and subjected to a treatment in order to knock-down Pasilla. The four untreated and three treated RNAi samples were used in the analysis. The treated samples had Pasilla knocked down by approximately 60% compared to the untreated samples. Some of the samples had undergone paired end sequencing while other samples were sequenced from one end only.</p>
<p>The reads were aligned to the Drosophila reference genome, downloaded from Ensembl, using the tophat aligner. The reads were summarised at the gene-level using htseq-count, a function from the tool HTSeq (<a href="http://wwwhuber.embl.de/users/anders/HTSeq/doc/overview.html" class="uri">http://wwwhuber.embl.de/users/anders/HTSeq/doc/overview.html</a>).</p>
<p>For the purpose of today’s workshop, we will be analysing the gene level counts.</p>
</div>
<div id="importing-the-data-into-r" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-the-data-into-r" class="anchor"></a>Importing the data into R</h2>
<p>First, let’s load all the libraries we will need today.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma">limma</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/talgalili/gplots">gplots</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Dm.eg.db</span><span class="op">)</span></code></pre></div>
<p>Next, read in the data and sample information file:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"https://ndownloader.figshare.com/files/5999835?private_link=e08e71c42f118dbe8be6"</span><span class="op">)</span>
<span class="va">sampleinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"https://ndownloader.figshare.com/files/5999838?private_link=e08e71c42f118dbe8be6"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Check that the data has benn imported correctly.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></code></pre></div>
<pre><code>##             SRR031714.bam SRR031716.bam SRR031724.bam SRR031726.bam
## FBgn0037213           157           142           213           291
## FBgn0000500             0             3             4             5
## FBgn0053294            14            18            13            19
## FBgn0037215          1666          1948          1662          1833
## FBgn0037217            13            19            18            23
## FBgn0037218           732           755           803           906
##             SRR031708.bam SRR031718.bam SRR031728.bam
## FBgn0037213           123           225           164
## FBgn0000500             2             1             1
## FBgn0053294            22            34            22
## FBgn0037215          1625          1701          1870
## FBgn0037217            15            15            17
## FBgn0037218           695           767           793</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sampleinfo</span></code></pre></div>
<pre><code>##   SampleName     Group Library
## 1  SRR031714 Untreated      PE
## 2  SRR031716 Untreated      PE
## 3  SRR031724   Treated      PE
## 4  SRR031726   Treated      PE
## 5  SRR031708 Untreated      SE
## 6  SRR031718   Treated      SE
## 7  SRR031728 Untreated      SE</code></pre>
</div>
<div id="filtering-out-lowly-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-out-lowly-expressed-genes" class="anchor"></a>Filtering out lowly expressed genes</h2>
<p>Our main interest is in testing the treated versus untreated groups. To check how many samples we have in each group we can use the <code>table</code> command.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##   Treated Untreated 
##         3         4</code></pre>
<p>The minimum sample size is 3. Let’s check the relationship between CPM and counts to see what CPM threshold we should be imposing. Recall we’re looking for a CPM that corresponds to a count of roughly 10-15.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mycpm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="va">mycpm</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">20</span><span class="op">)</span>,ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fl">10</span>,col<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">2</span>,col<span class="op">=</span><span class="fl">4</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>We can filter on a CPM of 2 or 3 in at least 3 samples.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">thresh</span> <span class="op">&lt;-</span> <span class="va">mycpm</span> <span class="op">&gt;</span> <span class="fl">2</span>
<span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">thresh</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></code></pre></div>
<pre><code>## keep
## FALSE  TRUE 
##  7345  7524</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts.keep</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">counts.keep</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 7524    7</code></pre>
<p>We are filtering out about half our genes.</p>
</div>
<div id="convert-to-dgelist-object" class="section level2">
<h2 class="hasAnchor">
<a href="#convert-to-dgelist-object" class="anchor"></a>Convert to DGEList object</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span><span class="va">counts.keep</span><span class="op">)</span></code></pre></div>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h2>
<p>Let’s do a number of quality control plots.</p>
<p>First, check the library sizes:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">lib.size</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Next check the distribution of the counts using a boxplot:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Get log2 counts per million</span>
<span class="va">logcpm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">counts</span>,log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="co"># Check distributions of samples using boxplots</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">logcpm</span>, xlab<span class="op">=</span><span class="st">""</span>, ylab<span class="op">=</span><span class="st">"Log2 counts per million"</span>,las<span class="op">=</span><span class="fl">2</span>,outline<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co"># Let's add a blue horizontal line that corresponds to the median logCPM</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">logcpm</span><span class="op">)</span>,col<span class="op">=</span><span class="st">"blue"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"Boxplots of logCPMs (unnormalised)"</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>We can colour by our groups, or by the different library prep.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,oma<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="va">group.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="st">"blue"</span><span class="op">)</span><span class="op">[</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Group</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">logcpm</span>, xlab<span class="op">=</span><span class="st">""</span>, ylab<span class="op">=</span><span class="st">"Log2 counts per million"</span>,las<span class="op">=</span><span class="fl">2</span>,col<span class="op">=</span><span class="va">group.col</span>,
        pars<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>cex.lab<span class="op">=</span><span class="fl">0.8</span>,cex.axis<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">logcpm</span><span class="op">)</span>,col<span class="op">=</span><span class="st">"blue"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"Boxplots of logCPMs\n(coloured by groups)"</span>,cex.main<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span>

<span class="va">lib.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"light pink"</span>,<span class="st">"light green"</span><span class="op">)</span><span class="op">[</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Library</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">logcpm</span>, xlab<span class="op">=</span><span class="st">""</span>, ylab<span class="op">=</span><span class="st">"Log2 counts per million"</span>,las<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="va">lib.col</span>,
        pars<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>cex.lab<span class="op">=</span><span class="fl">0.8</span>,cex.axis<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">logcpm</span><span class="op">)</span>,col<span class="op">=</span><span class="st">"blue"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"Boxplots of logCPMs\n(coloured by library prep)"</span>,cex.main<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>It doesn’t look like there are any large biases in the data.</p>
<p>Finally, let’s check our MDS plots.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html">plotMDS</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html">plotMDS</a></span><span class="op">(</span><span class="va">y</span>,col<span class="op">=</span><span class="va">group.col</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>,legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span>,fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="st">"blue"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html">plotMDS</a></span><span class="op">(</span><span class="va">y</span>,col<span class="op">=</span><span class="va">lib.col</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>,legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Library</span><span class="op">)</span>,fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"light pink"</span>,<span class="st">"light green"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>It looks like there is some variability in the data due to library type e.g. single end or paired end.</p>
</div>
<div id="hierarchical-clustering-with-heatmap-2" class="section level2">
<h2 class="hasAnchor">
<a href="#hierarchical-clustering-with-heatmap-2" class="anchor"></a>Hierarchical clustering with heatmap.2</h2>
<p>First we need a matrix of log counts:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">y</span>,log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Get variances for genes:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span></code></pre></div>
<p>Get top 500 most variable</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">select_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">var_genes</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">highly_variable_lcpm</span> <span class="op">&lt;-</span> <span class="va">logcounts</span><span class="op">[</span><span class="va">select_var</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">highly_variable_lcpm</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 500   7</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mypalette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">11</span>,<span class="st">"RdYlBu"</span><span class="op">)</span>
<span class="va">morecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span><span class="va">mypalette</span><span class="op">)</span>
<span class="co"># Plot the heatmap</span>
<span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span><span class="op">(</span><span class="va">highly_variable_lcpm</span>,col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu">morecols</span><span class="op">(</span><span class="fl">50</span><span class="op">)</span><span class="op">)</span>,trace<span class="op">=</span><span class="st">"none"</span>, main<span class="op">=</span><span class="st">"Top 500 most variable genes across samples"</span>,ColSideColors<span class="op">=</span><span class="va">group.col</span>,scale<span class="op">=</span><span class="st">"row"</span>,margins<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span><span class="op">(</span><span class="va">highly_variable_lcpm</span>,col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu">morecols</span><span class="op">(</span><span class="fl">50</span><span class="op">)</span><span class="op">)</span>,trace<span class="op">=</span><span class="st">"none"</span>, main<span class="op">=</span><span class="st">"Top 500 most variable genes across samples"</span>,ColSideColors<span class="op">=</span><span class="va">lib.col</span>,scale<span class="op">=</span><span class="st">"row"</span>,margins<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-17-2.png" width="700"></p>
</div>
<div id="normalisation" class="section level2">
<h2 class="hasAnchor">
<a href="#normalisation" class="anchor"></a>Normalisation</h2>
<p>Let’s do TMM normalisation</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>
<span class="va">y</span><span class="op">$</span><span class="va">samples</span></code></pre></div>
<pre><code>##               group lib.size norm.factors
## SRR031714.bam     1  4198181    0.9805029
## SRR031716.bam     1  4740969    0.9566649
## SRR031724.bam     1  4601291    0.9936278
## SRR031726.bam     1  5195516    1.0019699
## SRR031708.bam     1  3980986    1.0036556
## SRR031718.bam     1  4320527    1.0556860
## SRR031728.bam     1  4031608    1.0106328</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html">plotMD</a></span><span class="op">(</span><span class="va">logcounts</span>,column<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>,col<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html">plotMD</a></span><span class="op">(</span><span class="va">y</span>,column <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>,col<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div id="differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h2>
<div id="set-up-design-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#set-up-design-matrix" class="anchor"></a>Set up design matrix</h3>
<p>We want to test for differences between the treated and untreated samples. However, we know that the library preparation adds variability to the data, so we need to account for it in our model. We do this by modelling both Group and Library as variables in our design matrix. This is known as an additive model.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Library</span> <span class="op">+</span> <span class="va">sampleinfo</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span>
<span class="va">design</span></code></pre></div>
<pre><code>##   (Intercept) sampleinfo$LibrarySE sampleinfo$GroupUntreated
## 1           1                    0                         1
## 2           1                    0                         1
## 3           1                    0                         0
## 4           1                    0                         0
## 5           1                    1                         1
## 6           1                    1                         0
## 7           1                    1                         1
## attr(,"assign")
## [1] 0 1 2
## attr(,"contrasts")
## attr(,"contrasts")$`sampleinfo$Library`
## [1] "contr.treatment"
## 
## attr(,"contrasts")$`sampleinfo$Group`
## [1] "contr.treatment"</code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Int"</span>,<span class="st">"SEvsPE"</span>,<span class="st">"UVsT"</span><span class="op">)</span></code></pre></div>
</div>
<div id="voom-transform-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#voom-transform-the-data" class="anchor"></a>Voom transform the data</h3>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html">voom</a></span><span class="op">(</span><span class="va">y</span>,<span class="va">design</span>,plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span>,col<span class="op">=</span><span class="fl">4</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">v</span><span class="op">$</span><span class="va">E</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">v</span><span class="op">$</span><span class="va">E</span><span class="op">)</span>,col<span class="op">=</span><span class="fl">4</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div id="test-for-differential-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#test-for-differential-expression" class="anchor"></a>Test for differential expression</h3>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="va">v</span>,<span class="va">design</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span>
<span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html">decideTests</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/NMF/man/assess.html">summary</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></code></pre></div>
<pre><code>##         Int SEvsPE UVsT
## Down      4    125  440
## NotSig   66   7142 6641
## Up     7454    257  443</code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>,coef<span class="op">=</span><span class="fl">3</span>,sort.by<span class="op">=</span><span class="st">"p"</span><span class="op">)</span></code></pre></div>
<pre><code>##                 logFC   AveExpr         t      P.Value    adj.P.Val        B
## FBgn0025111 -2.863110  6.253448 -31.24326 9.756044e-13 4.734428e-09 19.04663
## FBgn0003360  3.089242  7.778378  30.57062 1.258487e-12 4.734428e-09 19.04370
## FBgn0026562  2.422219 11.359837  24.25592 1.868075e-11 4.685133e-08 16.57869
## FBgn0029167  2.247086  7.825170  21.77945 6.509505e-11 1.224438e-07 15.57725
## FBgn0035085  2.677008  5.180607  18.97509 3.188148e-10 4.797524e-07 13.66077
## FBgn0039155  4.272254  4.800907  17.42222 8.476440e-10 1.062946e-06 12.15024
## FBgn0040091  1.559497  6.245766  15.43293 3.363833e-09 3.615639e-06 11.72125
## FBgn0023479  1.576335  7.768886  15.06473 4.419245e-09 3.877495e-06 11.43434
## FBgn0029896  2.312097  4.881577  15.00033 4.638152e-09 3.877495e-06 11.25267
## FBgn0027279  1.180272  7.760889  14.37303 7.503535e-09 5.645660e-06 10.89435</code></pre>
</div>
<div id="add-annotation-from-org-dm-eg-db" class="section level3">
<h3 class="hasAnchor">
<a href="#add-annotation-from-org-dm-eg-db" class="anchor"></a>Add annotation from org.Dm.eg.db</h3>
<p>The rownames of the fit object are <a href="FlyBase">http://flybase.org/</a> ids.</p>
<p>First we need to decide what information we want. In order to see what we can extract we can run the <code>columns</code> function on the annotation database.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">columns</span><span class="op">(</span><span class="va">org.Dm.eg.db</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "FLYBASE"     
## [11] "FLYBASECG"    "FLYBASEPROT"  "GENENAME"     "GO"           "GOALL"       
## [16] "MAP"          "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PMID"        
## [21] "REFSEQ"       "SYMBOL"       "UNIGENE"      "UNIPROT"</code></pre>
<p>We definitely want gene symbols and perhaps the full gene name and Entrez id. Let’s build up our annotation information in a separate data frame using the <code>select</code> function. Note, by default, the <code>select</code> function assumes that the keys provided are Entrez ids. However, in this case we are using FlyBase ids as the keys. As such, we need to give the <code>select</code> function this information using the <code>keytype</code> argument which we will set to <code>"FLYBASE"</code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ann</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">org.Dm.eg.db</span>,keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span>,columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FLYBASE"</span>,<span class="st">"ENTREZID"</span>,<span class="st">"SYMBOL"</span>,<span class="st">"GENENAME"</span><span class="op">)</span>,keytype<span class="op">=</span><span class="st">"FLYBASE"</span><span class="op">)</span></code></pre></div>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Have a look at the annotation</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ann</span><span class="op">)</span></code></pre></div>
<pre><code>##       FLYBASE ENTREZID   SYMBOL                GENENAME
## 1 FBgn0037213    40522  CG12581 uncharacterized protein
## 2 FBgn0053294  3772637  CR33294                  pseudo
## 3 FBgn0037215    40524 beta-Man        beta-Mannosidase
## 4 FBgn0037217    40526  CG14636 uncharacterized protein
## 5 FBgn0037218    40527      aux                 auxilin
## 6 FBgn0261436    40528     DhpD Dihydropterin deaminase</code></pre>
<p>Let’s double check that the <code>FLYBASE</code> column matches exactly to our <code>fit</code> rownames.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ann</span><span class="op">$</span><span class="va">FLYBASE</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## TRUE 
## 7524</code></pre>
<p>We can slot in the annotation information into the <code>genes</code> slot of <code>fit</code>. (Please note that if the <code>select</code> function returns a 1:many mapping then you can’t just append the annotation to the fit object. An alternative way to get annotation is shown below.)</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fit</span><span class="op">$</span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">ann</span></code></pre></div>
<p>Now when we run the <code>topTable</code> command, the annotation information should be included in the output.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>,coef<span class="op">=</span><span class="fl">3</span>,sort.by<span class="op">=</span><span class="st">"p"</span><span class="op">)</span></code></pre></div>
<pre><code>##                 FLYBASE ENTREZID    SYMBOL
## FBgn0025111 FBgn0025111    32008      Ant2
## FBgn0003360 FBgn0003360    32007      sesB
## FBgn0026562 FBgn0026562    43230     SPARC
## FBgn0029167 FBgn0029167    39529       Hml
## FBgn0035085 FBgn0035085    37991    CG3770
## FBgn0039155 FBgn0039155    42865      Kal1
## FBgn0040091 FBgn0040091    37590  Ugt317A1
## FBgn0023479 FBgn0023479    39048       teq
## FBgn0029896 FBgn0029896    31612    CG3168
## FBgn0027279 FBgn0027279    33137 l(1)G0196
##                                                 GENENAME     logFC   AveExpr
## FBgn0025111             Adenine nucleotide translocase 2 -2.863110  6.253448
## FBgn0003360                           stress-sensitive B  3.089242  7.778378
## FBgn0026562      Secreted protein, acidic, cysteine-rich  2.422219 11.359837
## FBgn0029167                                   Hemolectin  2.247086  7.825170
## FBgn0035085                      uncharacterized protein  2.677008  5.180607
## FBgn0039155                          Kallmann syndrome 1  4.272254  4.800907
## FBgn0040091 UDP-glycosyltransferase family 317 member A1  1.559497  6.245766
## FBgn0023479                                      Tequila  1.576335  7.768886
## FBgn0029896                      uncharacterized protein  2.312097  4.881577
## FBgn0027279                             lethal (1) G0196  1.180272  7.760889
##                     t      P.Value    adj.P.Val        B
## FBgn0025111 -31.24326 9.756044e-13 4.734428e-09 19.04663
## FBgn0003360  30.57062 1.258487e-12 4.734428e-09 19.04370
## FBgn0026562  24.25592 1.868075e-11 4.685133e-08 16.57869
## FBgn0029167  21.77945 6.509505e-11 1.224438e-07 15.57725
## FBgn0035085  18.97509 3.188148e-10 4.797524e-07 13.66077
## FBgn0039155  17.42222 8.476440e-10 1.062946e-06 12.15024
## FBgn0040091  15.43293 3.363833e-09 3.615639e-06 11.72125
## FBgn0023479  15.06473 4.419245e-09 3.877495e-06 11.43434
## FBgn0029896  15.00033 4.638152e-09 3.877495e-06 11.25267
## FBgn0027279  14.37303 7.503535e-09 5.645660e-06 10.89435</code></pre>
<p>If for some reason the <code>select</code> function doesn’t work, or the result is a 1:many mapping, there is a less elegant way to get the annotations using the <code>toTable</code> command.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's see what we can get in table format from org.Dm.eg.db</span>
<span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="st">"package:org.Dm.eg.db"</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "org.Dm.eg"                "org.Dm.eg_dbconn"        
##  [3] "org.Dm.eg_dbfile"         "org.Dm.eg_dbInfo"        
##  [5] "org.Dm.eg_dbschema"       "org.Dm.eg.db"            
##  [7] "org.Dm.egACCNUM"          "org.Dm.egACCNUM2EG"      
##  [9] "org.Dm.egALIAS2EG"        "org.Dm.egCHR"            
## [11] "org.Dm.egCHRLENGTHS"      "org.Dm.egCHRLOC"         
## [13] "org.Dm.egCHRLOCEND"       "org.Dm.egENSEMBL"        
## [15] "org.Dm.egENSEMBL2EG"      "org.Dm.egENSEMBLPROT"    
## [17] "org.Dm.egENSEMBLPROT2EG"  "org.Dm.egENSEMBLTRANS"   
## [19] "org.Dm.egENSEMBLTRANS2EG" "org.Dm.egENZYME"         
## [21] "org.Dm.egENZYME2EG"       "org.Dm.egFLYBASE"        
## [23] "org.Dm.egFLYBASE2EG"      "org.Dm.egFLYBASECG"      
## [25] "org.Dm.egFLYBASECG2EG"    "org.Dm.egFLYBASEPROT"    
## [27] "org.Dm.egFLYBASEPROT2EG"  "org.Dm.egGENENAME"       
## [29] "org.Dm.egGO"              "org.Dm.egGO2ALLEGS"      
## [31] "org.Dm.egGO2EG"           "org.Dm.egMAP"            
## [33] "org.Dm.egMAP2EG"          "org.Dm.egMAPCOUNTS"      
## [35] "org.Dm.egORGANISM"        "org.Dm.egPATH"           
## [37] "org.Dm.egPATH2EG"         "org.Dm.egPMID"           
## [39] "org.Dm.egPMID2EG"         "org.Dm.egREFSEQ"         
## [41] "org.Dm.egREFSEQ2EG"       "org.Dm.egSYMBOL"         
## [43] "org.Dm.egSYMBOL2EG"       "org.Dm.egUNIGENE"        
## [45] "org.Dm.egUNIGENE2EG"      "org.Dm.egUNIPROT"</code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get annotation</span>
<span class="va">fly</span> <span class="op">&lt;-</span> <span class="fu">toTable</span><span class="op">(</span><span class="va">org.Dm.egFLYBASE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">fly</span><span class="op">)</span></code></pre></div>
<pre><code>##   gene_id  flybase_id
## 1   30970 FBgn0040373
## 2   30971 FBgn0040372
## 3   30972 FBgn0261446
## 4   30973 FBgn0000316
## 5   30975 FBgn0005427
## 6   30976 FBgn0040370</code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">symbol</span> <span class="op">&lt;-</span> <span class="fu">toTable</span><span class="op">(</span><span class="va">org.Dm.egSYMBOL</span><span class="op">)</span>
<span class="va">genename</span> <span class="op">&lt;-</span> <span class="fu">toTable</span><span class="op">(</span><span class="va">org.Dm.egGENENAME</span><span class="op">)</span>

<span class="co"># We can use the merge command to merge two dataframes</span>
<span class="va">ann1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">fly</span>,<span class="va">symbol</span>,by<span class="op">=</span><span class="st">"gene_id"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ann1</span><span class="op">)</span></code></pre></div>
<pre><code>##    gene_id  flybase_id        symbol
## 1 10178776 FBgn0261702 asRNA:CR42738
## 2 10178777 FBgn0262024       CG42835
## 3 10178779 FBgn0058354       CR40354
## 4 10178780 FBgn0053929       CR33929
## 5 10178781 FBgn0262141       CG42867
## 6 10178782 FBgn0062978       CG31808</code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Add genename table to ann1</span>
<span class="va">ann2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">ann1</span>,<span class="va">genename</span>,by<span class="op">=</span><span class="st">"gene_id"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ann2</span><span class="op">)</span></code></pre></div>
<pre><code>##    gene_id  flybase_id        symbol               gene_name
## 1 10178776 FBgn0261702 asRNA:CR42738   antisense RNA:CR42738
## 2 10178777 FBgn0262024       CG42835 uncharacterized protein
## 3 10178779 FBgn0058354       CR40354                  pseudo
## 4 10178780 FBgn0053929       CR33929                  pseudo
## 5 10178781 FBgn0262141       CG42867 uncharacterized protein
## 6 10178782 FBgn0062978       CG31808 uncharacterized protein</code></pre>
<p>Now we need to match up between <code><a href="https://rdrr.io/r/base/colnames.html">rownames(fit)</a></code> and the ensemble gene id in <code>ann2</code>.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span>,<span class="va">ann2</span><span class="op">$</span><span class="va">flybase_id</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span> <span class="co"># check for unmatched rows</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##  6911   613</code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ann3</span> <span class="op">&lt;-</span> <span class="va">ann2</span><span class="op">[</span><span class="va">m</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">]</span>,<span class="op">]</span> <span class="co"># exclude unmatched rows</span>

<span class="co"># compare the results in ann3 to what is in the fit object</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ann3</span><span class="op">)</span></code></pre></div>
<pre><code>##       gene_id  flybase_id   symbol               gene_name
## 16229   40522 FBgn0037213  CG12581 uncharacterized protein
## 13723 3772637 FBgn0053294  CR33294                  pseudo
## 16230   40524 FBgn0037215 beta-Man        beta-Mannosidase
## 16231   40526 FBgn0037217  CG14636 uncharacterized protein
## 16232   40527 FBgn0037218      aux                 auxilin
## 16233   40528 FBgn0261436     DhpD Dihydropterin deaminase</code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span></code></pre></div>
<pre><code>##       FLYBASE ENTREZID   SYMBOL                GENENAME
## 1 FBgn0037213    40522  CG12581 uncharacterized protein
## 2 FBgn0053294  3772637  CR33294                  pseudo
## 3 FBgn0037215    40524 beta-Man        beta-Mannosidase
## 4 FBgn0037217    40526  CG14636 uncharacterized protein
## 5 FBgn0037218    40527      aux                 auxilin
## 6 FBgn0261436    40528     DhpD Dihydropterin deaminase</code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>,coef<span class="op">=</span><span class="fl">3</span>,sort.by<span class="op">=</span><span class="st">"p"</span><span class="op">)</span></code></pre></div>
<pre><code>##                 FLYBASE ENTREZID    SYMBOL
## FBgn0025111 FBgn0025111    32008      Ant2
## FBgn0003360 FBgn0003360    32007      sesB
## FBgn0026562 FBgn0026562    43230     SPARC
## FBgn0029167 FBgn0029167    39529       Hml
## FBgn0035085 FBgn0035085    37991    CG3770
## FBgn0039155 FBgn0039155    42865      Kal1
## FBgn0040091 FBgn0040091    37590  Ugt317A1
## FBgn0023479 FBgn0023479    39048       teq
## FBgn0029896 FBgn0029896    31612    CG3168
## FBgn0027279 FBgn0027279    33137 l(1)G0196
##                                                 GENENAME     logFC   AveExpr
## FBgn0025111             Adenine nucleotide translocase 2 -2.863110  6.253448
## FBgn0003360                           stress-sensitive B  3.089242  7.778378
## FBgn0026562      Secreted protein, acidic, cysteine-rich  2.422219 11.359837
## FBgn0029167                                   Hemolectin  2.247086  7.825170
## FBgn0035085                      uncharacterized protein  2.677008  5.180607
## FBgn0039155                          Kallmann syndrome 1  4.272254  4.800907
## FBgn0040091 UDP-glycosyltransferase family 317 member A1  1.559497  6.245766
## FBgn0023479                                      Tequila  1.576335  7.768886
## FBgn0029896                      uncharacterized protein  2.312097  4.881577
## FBgn0027279                             lethal (1) G0196  1.180272  7.760889
##                     t      P.Value    adj.P.Val        B
## FBgn0025111 -31.24326 9.756044e-13 4.734428e-09 19.04663
## FBgn0003360  30.57062 1.258487e-12 4.734428e-09 19.04370
## FBgn0026562  24.25592 1.868075e-11 4.685133e-08 16.57869
## FBgn0029167  21.77945 6.509505e-11 1.224438e-07 15.57725
## FBgn0035085  18.97509 3.188148e-10 4.797524e-07 13.66077
## FBgn0039155  17.42222 8.476440e-10 1.062946e-06 12.15024
## FBgn0040091  15.43293 3.363833e-09 3.615639e-06 11.72125
## FBgn0023479  15.06473 4.419245e-09 3.877495e-06 11.43434
## FBgn0029896  15.00033 4.638152e-09 3.877495e-06 11.25267
## FBgn0027279  14.37303 7.503535e-09 5.645660e-06 10.89435</code></pre>
<p>Let’s check the expression of pasilla.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"pasilla"</span>,<span class="va">fit</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">GENENAME</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span><span class="op">[</span><span class="va">ps</span>,<span class="op">]</span>,coef<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##                 FLYBASE ENTREZID SYMBOL GENENAME    logFC  AveExpr       t
## FBgn0261552 FBgn0261552    44258     ps  pasilla 1.845368 8.344996 11.5651
##                 P.Value   adj.P.Val        B
## FBgn0261552 8.37355e-08 8.37355e-08 8.425095</code></pre>
</div>
</div>
<div id="plots-after-testing-for-de" class="section level2">
<h2 class="hasAnchor">
<a href="#plots-after-testing-for-de" class="anchor"></a>Plots after testing for DE</h2>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html">plotMD</a></span><span class="op">(</span><span class="va">fit</span>,coef<span class="op">=</span><span class="fl">3</span>,status<span class="op">=</span><span class="va">results</span><span class="op">[</span>,<span class="st">"UVsT"</span><span class="op">]</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/limma/man/volcanoplot.html">volcanoplot</a></span><span class="op">(</span><span class="va">fit</span>,coef<span class="op">=</span><span class="fl">3</span>,highlight<span class="op">=</span><span class="fl">100</span>,names<span class="op">=</span><span class="va">fit</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/stripchart.html">stripchart</a></span><span class="op">(</span><span class="va">v</span><span class="op">$</span><span class="va">E</span><span class="op">[</span><span class="st">"FBgn0025111"</span>,<span class="op">]</span><span class="op">~</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Check expression of Pasilla</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/stripchart.html">stripchart</a></span><span class="op">(</span><span class="va">v</span><span class="op">$</span><span class="va">E</span><span class="op">[</span><span class="st">"FBgn0261552"</span>,<span class="op">]</span><span class="op">~</span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
</div>
<div id="testing-relative-to-a-threshold" class="section level2">
<h2 class="hasAnchor">
<a href="#testing-relative-to-a-threshold" class="anchor"></a>Testing relative to a threshold</h2>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fit.treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">treat</a></span><span class="op">(</span><span class="va">fit</span>,lfc<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="va">res.treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html">decideTests</a></span><span class="op">(</span><span class="va">fit.treat</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/NMF/man/assess.html">summary</a></span><span class="op">(</span><span class="va">res.treat</span><span class="op">)</span></code></pre></div>
<pre><code>##         Int SEvsPE UVsT
## Down      2     13   10
## NotSig  350   7510 7496
## Up     7172      1   18</code></pre>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTreat</a></span><span class="op">(</span><span class="va">fit.treat</span>,coef<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##                 FLYBASE ENTREZID SYMBOL                                GENENAME
## FBgn0003360 FBgn0003360    32007   sesB                      stress-sensitive B
## FBgn0025111 FBgn0025111    32008   Ant2        Adenine nucleotide translocase 2
## FBgn0026562 FBgn0026562    43230  SPARC Secreted protein, acidic, cysteine-rich
## FBgn0039155 FBgn0039155    42865   Kal1                     Kallmann syndrome 1
## FBgn0029167 FBgn0029167    39529    Hml                              Hemolectin
## FBgn0035085 FBgn0035085    37991 CG3770                 uncharacterized protein
## FBgn0034736 FBgn0034736    37572    gas                                gasoline
## FBgn0000071 FBgn0000071    40831    Ama                                 Amalgam
## FBgn0029896 FBgn0029896    31612 CG3168                 uncharacterized protein
## FBgn0039827 FBgn0039827    43689 CG1544                 uncharacterized protein
##                 logFC   AveExpr          t      P.Value    adj.P.Val
## FBgn0003360  3.089242  7.778378  20.674784 5.940461e-11 2.711163e-07
## FBgn0025111 -2.863110  6.253448 -20.330911 7.206708e-11 2.711163e-07
## FBgn0026562  2.422219 11.359837  14.241997 4.158446e-09 1.042938e-05
## FBgn0039155  4.272254  4.800907  13.344224 8.653631e-09 1.627748e-05
## FBgn0029167  2.247086  7.825170  12.087142 2.578989e-08 3.880863e-05
## FBgn0035085  2.677008  5.180607  11.886920 3.099362e-08 3.886600e-05
## FBgn0034736  3.407306  3.372966   8.780952 7.931358e-07 8.094637e-04
## FBgn0000071 -2.572434  4.209333  -8.710932 8.606738e-07 8.094637e-04
## FBgn0029896  2.312097  4.881577   8.512571 1.091548e-06 8.898178e-04
## FBgn0039827  3.925533  3.204658   8.449359 1.182639e-06 8.898178e-04</code></pre>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html">plotMD</a></span><span class="op">(</span><span class="va">fit.treat</span>,coef<span class="op">=</span><span class="fl">3</span>,status<span class="op">=</span><span class="va">res.treat</span><span class="op">[</span>,<span class="st">"UVsT"</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>,col<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></code></pre></div>
<p><img src="09-applying-rnaseq-solutions_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
</div>
<div id="gene-set-testing-with-goana" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-set-testing-with-goana" class="anchor"></a>Gene set testing with goana</h2>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">go</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/goana.html">goana</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="st">"UVsT"</span>,species <span class="op">=</span> <span class="st">"Dm"</span>, geneid<span class="op">=</span><span class="st">"ENTREZID"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/limma/man/topGO.html">topGO</a></span><span class="op">(</span><span class="va">go</span>, n<span class="op">=</span><span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>##                                                  Term Ont   N Up Down
## GO:0019991                  septate junction assembly  BP  29  0   12
## GO:0120192                    tight junction assembly  BP  31  0   12
## GO:0120193                tight junction organization  BP  31  0   12
## GO:0005576                       extracellular region  CC 318 43   43
## GO:0043297                   apical junction assembly  BP  36  0   12
## GO:0005615                        extracellular space  CC 181 29   29
## GO:0045216            cell-cell junction organization  BP  59  0   15
## GO:0006006                  glucose metabolic process  BP  22  2    9
## GO:0060856       establishment of blood-brain barrier  BP  17  2    8
## GO:0060857 establishment of glial blood-brain barrier  BP  13  0    7
##                    P.Up       P.Down
## GO:0019991 1.000000e+00 2.835702e-08
## GO:0120192 1.000000e+00 6.926354e-08
## GO:0120193 1.000000e+00 6.926354e-08
## GO:0005576 1.260580e-07 1.453207e-07
## GO:0043297 1.000000e+00 4.698711e-07
## GO:0005615 4.764942e-07 5.288276e-07
## GO:0045216 1.000000e+00 9.120552e-07
## GO:0006006 3.690371e-01 1.835789e-06
## GO:0060856 2.598739e-01 1.940334e-06
## GO:0060857 1.000000e+00 2.797378e-06</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
